testsuite/tests/typing-abstract-kinds/basics.ml

```diff
@@ -0,0 +1,766 @@
+|}]
+
+kind_ k_portable = k mod portable
+
+type t_portable : k_portable
+type s_portable : any mod portable = t_portable
+[%%expect{|
+kind_ k_portable = k mod portable
+type t_portable : k mod portable
+type s_portable = t_portable
+|}]
+
+(****************************************************)
+(* Test: kind aliases work like the kind they alias *)
+
+kind_ k = value & value mod portable
+
+type t : k
+
+type s1 : value & value = t
+type s2 : value & value mod portable = t
+type s3 : any mod portable = t
+type s4 : any = t
+
+let require_portable (_x : t @ portable) = ()
+let cross_portable : t -> unit = fun x -> require_portable x
+[%%expect{|
+kind_ k = value mod portable & value mod portable
+type t : value mod portable & value mod portable
+type s1 = t
+type s2 = t
+type s3 = t
+type s4 = t
+val require_portable : t @ portable -> unit = <fun>
+val cross_portable : t -> unit = <fun>
+|}]
+
+type s5 : value & value mod global = t
+[%%expect{|
+Line 1, characters 0-38:
+1 | type s5 : value & value mod global = t
+    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Error: The kind of type "t" is value mod portable & value mod portable
+         because of the definition of t at line 3, characters 0-10.
+       But the kind of type "t" must be a subkind of
+           value mod global & value mod global
+         because of the definition of s5 at line 1, characters 0-38.
+|}]
+
+let does_not_cross_local (x : t @ local) : t @ global = x
+[%%expect{|
+Line 1, characters 56-57:
+1 | let does_not_cross_local (x : t @ local) : t @ global = x
+                                                            ^
+Error: This value is "local" to the parent region but is expected to be "global".
+|}]
+
+(****************************************************************)
+(* Test: expanding through aliases takes the meet of the bounds *)
+
+module A = struct
+  kind_ ka = float64
+end
+
+module B = struct
+  kind_ kb = A.ka mod portable
+end
+
+module C = struct
+  kind_ kc = B.kb
+end
+
+module D = struct
+  kind_ kd = C.kc mod global
+end
+
+type t : D.kd
+
+type s1 : float64 mod portable global = t
+type s2 : any mod portable global = t
+[%%expect{|
+module A : sig kind_ ka = float64 end
+module B : sig kind_ kb = float64 mod portable end
+module C : sig kind_ kc = float64 mod portable end
+module D : sig kind_ kd = float64 mod global portable end
+type t : float64 mod global portable
+type s1 = t
+type s2 = t
+|}]
+
+type s3 : any mod contended = t
+[%%expect{|
+Line 1, characters 0-31:
+1 | type s3 : any mod contended = t
+    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Error: The kind of type "t" is float64 mod global portable
+         because of the definition of t at line 17, characters 0-13.
+       But the kind of type "t" must be a subkind of any mod contended
+         because of the definition of s3 at line 1, characters 0-31.
+|}]
+
+(************************)
+(* Test: no with bounds *)
+
+type t
+
+kind_ k = immutable_data with t
```

Comment:
Might be nice to also look about non-abstract with-bounds. Like

```
kind_ k = immutable_data with int
```

This actually is fine -- we can just expand away the `int` -- but I think it's simpler to reject.